txnsync/encodedgroupsmarshalers.go

```diff
@@ -0,0 +1,826 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package txnsync
+
+import (
+	"fmt"
+
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+func squeezeByteArray(b []byte) []byte {
+	if len(b)%2 == 1 {
+		b = append(b, byte(0))
+	}
+	for index := 0; index*2 < len(b); index++ {
+		b[index] = b[index*2]*16 + b[index*2+1]
+	}
+	return b[0 : len(b)/2]
+}
+
+func addGroupHashes(txnGroups []transactions.SignedTxGroup, txnCount int, b bitmask) {
+	index := 0
+	txGroupHashes := make([]crypto.Digest, txnCount)
+	for _, txns := range txnGroups {
+		var txGroup transactions.TxGroup
+		txGroup.TxGroupHashes = txGroupHashes[index : index+len(txns.Transactions)]
+		for i, tx := range txns.Transactions {
+			txGroup.TxGroupHashes[i] = crypto.HashObj(tx.Txn)
+		}
+		groupHash := crypto.HashObj(txGroup)
+		for i := range txns.Transactions {
+			if exists := b.EntryExists(index, txnCount); exists {
+				txns.Transactions[i].Txn.Group = groupHash
+			}
```

Comment:
The predicate here is too loose. The group hash need to be added if one of the following applies
1. this transaction group contains more than 1 transaction
2. the bitmask is set.